crates/volta-core/src/toolchain/mod.rs

```diff
@@ -95,39 +96,32 @@ impl Toolchain {
 
     /// Set the active Yarn version in the default platform file.
     pub fn set_active_yarn(&mut self, yarn_version: &Version) -> Fallible<()> {
-        let mut dirty = false;
-
-        if let Some(ref mut platform) = self.platform {
-            if platform.yarn.as_ref() != Some(yarn_version) {
-                platform.yarn = Some(yarn_version.clone());
-                dirty = true;
+        if let Some(platform) = &self.platform {
+            if platform.yarn.as_ref().map(|y| &y.version) != Some(yarn_version) {
+                self.platform = Some(Rc::new(PlatformSpec {
+                    node: platform.node.clone(),
+                    npm: platform.npm.clone(),
+                    yarn: Some(SourcedVersion::default(yarn_version.clone())),
+                }));
+                self.save()?;
```

Comment:
Reading these implementations, I see what you mean about repetition. I wonder if the ergonomic improvement might be to have methods on `PlatformSpec` like this:

```rust
impl PlatformSpec {
    fn with_new_default_yarn(&self, yarn_version: &Version) -> Rc<PlatformSpec> {
        Rc::new(PlatformSpec {
            node: self.node.clone(),
            npm: self.npm.clone(),
            yarn: Some(SourcedVersion::default(yarn_version.clone()),
        })
    }
}
```

Then in each of these places, you'd use it like:

```rust
if platform.yarn.as_ref().map(|y| &y.version) != Some(yarn_version) {
    self.platform = self.platform.with_new_default_yarn(yarn_version);
}
```

If you wanted, you could even go so far as to do something like this:

```rust
impl PlatformSpec {
    fn with_updated_yarn(&self, yarn_version: &VersionSpec, source: Fn(Version) -> Source) {
        Rc::new(PlatformSpec {
            node: self.node.clone(),
            npm: self.npm.clone(),
            yarn: Some(source(yarn_version.clone()),
        })
    }
}
```

That in turn would let you just do:

```rust
if platform.yarn.as_ref().map(|y| &y.version) != Some(yarn_version) {
    self.platform = self.platform.with_updated_yarn(yarn_version, SourcedVersion::default);
}
```

etc.

Reply:
I like that, and I was thinking that adding some convenience methods would help clean up a lot of the issues I felt when converting. I'll look at adding those in addition to the Trait idea I had and see if I can come up with something that feels great to use.